Adding READMEs to the repo (#270)
ros/diagnostics: diagnostic_aggregator/README.md

> # Overview

Reviewer: ```suggestion
General information about this repository, including legal information and known issues/limitations, are given in [README.md](../README.md) in the repository root.

# The diagnostic_aggregator package
```

This package contains the `aggregator_node`.
It listens to the [`diagnostic_msgs/DiagnosticArray`](https://index.ros.org/p/diagnostic_msgs) messages on the `/diagnostics` topic and aggregates and published them on the `/diagnostics_agg` topic.

## Example
One example of how to use this package is to aggregate the diagnostics of a robot.
Aggregation means that the diagnostics of the robot are grouped by various aspects, like their location on the robot, their type, etc.

> This will allow you to easily see which part of the robot is not working properly.

Reviewer: I propose to move this explanation to the previous section and to start the example section directly with 'In our example ...'

In our example, we are looking at a robot with arms and legs.
The robot has two of each, one on each side.
The robot also 4 camera sensors, one left and one right and one in the front and one in the back.
These are all the available diagnostic sources:

``` 
/arms/left/motor
/arms/right/motor
/legs/left/motor
/legs/right/motor
/sensors/left/cam
/sensors/right/cam
/sensors/front/cam
/sensors/rear/cam
```

We want to group the diagnostics by
- all sensors
- all motors
- left side of the robot
- right side of the robot

> We can achieve that by creating a configuration file that looks like this (see [example_analyzers.yaml](diagnostic_aggregator/example/example_analyzers.yaml)):

Reviewer: ```suggestion
We can achieve that by creating a configuration file that looks like this (see [example_analyzers.yaml](example/example_analyzers.yaml)):
```

``` yaml
analyzers:
  ros__parameters:
    path: Aggregation
    arms:
      type: diagnostic_aggregator/GenericAnalyzer
      path: Arms
      startswith: [ '/arms' ]
    legs:
      type: diagnostic_aggregator/GenericAnalyzer
      path: Legs
      startswith: [ '/legs' ]
    sensors:
      type: diagnostic_aggregator/GenericAnalyzer
      path: Sensors
      startswith: [ '/sensors' ]
    motors:
      type: diagnostic_aggregator/GenericAnalyzer
      path: Motors
      contains: [ '/motor' ]
    topology:
      type: 'diagnostic_aggregator/AnalyzerGroup'
      path: Topology
      analyzers:
        left:
          type: diagnostic_aggregator/GenericAnalyzer
          path: Left
          contains: [ '/left' ]
        right:
          type: diagnostic_aggregator/GenericAnalyzer
          path: Right
          contains: [ '/right' ]
```

> This will allow [rqt_robot_monitor](https://index.ros.org/p/rqt_robot_monitor) to display the diagnostics in a more useful way:

Reviewer: ```suggestion
Based on this configuration, the [rqt_robot_monitor](https://index.ros.org/p/rqt_robot_monitor) will display the diagnostics information in a well-arranged manner as follows:
```

![doc/rqt_robot_monitor.png](doc/rqt_robot_monitor.png)

Note that it will also display the highest state per group to allow you to see at a glance which part of the robot is not working properly.

> For example in the above image, the left side of the robot is not working properly, because the left cam is in the `ERROR` state.

Reviewer: ```suggestion
For example in the above image, the left side of the robot is not working properly, because the left camera is in the `ERROR` state.
```

# Analyzers
The `aggregator_node` will load analyzers to process the diagnostics data.

> An analyzer is a plugin that inherits from the [`diagnostic_aggregator::Analyzer`](diagnostic_aggregator/include/diagnostic_aggregator/analyzer.hpp) class.

Reviewer: ```suggestion
An analyzer is a plugin that inherits from the [`diagnostic_aggregator::Analyzer`](include/diagnostic_aggregator/analyzer.hpp) class.
```

Analyzers must be implemented in packages that directly depend on [`pluginlib`](https://index.ros.org/p/pluginlib) and `diagnostic_aggregator`.

> The [`diagnostic_aggregator::Analyzer`](diagnostic_aggregator/include/diagnostic_aggregator/analyzer.hpp) class is purely virtual and derived classes must implement the following methods:

Reviewer: ```suggestion
The [`diagnostic_aggregator::Analyzer`](include/diagnostic_aggregator/analyzer.hpp) class is purely virtual and derived classes must implement the following methods:
```

- `init()` - Analyzer is initialized with base path and namespace
- `match()` - Returns true if the analyzer is interested in the status message
- `analyze()` - Returns true if the analyzer will analyze the status message
- `report()` - Returns results of analysis as vector of status messages

> - `getPath()` - Returns the prefix path of the analyzer (e.g. "/robot/motors/", etc.)

Reviewer: ```suggestion
- `getPath()` - Returns the prefix path of the analyzer (e.g., "/robot/motors/")
```

> - `getName()` - Returns the name of the analyzer (e.g. "Motors", etc.)

Reviewer: ```suggestion
- `getName()` - Returns the name of the analyzer (e.g., "Motors")
```

Analyzers can choose the value of the error level of their output.
Usually, the error level of the output is the highest error level of the input.
The analyzers are responsible for setting the name of each item in the output correctly.

# Using the aggregator_node

## Configuration

The `aggregator_node` can be configured at launch time like in this example:
``` yaml
pub_rate: 1.0 # Optional, defaults to 1.0
base_path: 'PRE' # Optional, defaults to ""
analyzers:
  motors:
    type: PR2MotorsAnalyzer
  joints:
    type: GenericAnalyzer
    path: 'Joints'
    regex: 'Joint*'
```

The `pub_rate` parameter is the rate at which the aggregated diagnostics will be published.
The `base_path` parameter is the prefix that will be added to the name of each item in the output.

Under the `analyzers` key, you can specify the analyzers that you want to use.
Each analyzer must have a unique name.

> Under the name, you must specify the type of the analyzer, this must be the name of the class that implements the analyzer.

Reviewer: ```suggestion
Under the name, you must specify the type of the analyzer. This must be the name of the class that implements the analyzer.
```

``` python
    aggregator = launch_ros.actions.Node(
        package='diagnostic_aggregator',
        executable='aggregator_node',
        output='screen',
        parameters=[analyzer_params_filepath])
    return launch.LaunchDescription([
        aggregator,
    ])
```

# Basic analyzers
The `diagnostic_aggregator` package provides a few basic analyzers that you can use to aggregate your diagnostics.

## GenericAnalyzer
The [`diagnostic_aggregator::GenericAnalyzer`](include/diagnostic_aggregator/generic_analyzer.hpp) class is a basic analyzer that can be configured to match diagnostics based on their name.
By defining a `path` parameter, you can specify the prefix that will be added to the name of each item in the output.
This way you can group diagnostics by their location or other aspects as demonstrated in the [example](#example).

## AnalyzerGroup
The [`diagnostic_aggregator::AnalyzerGroup`](include/diagnostic_aggregator/analyzer_group.hpp) class is a basic analyzer that can be configured to group other analyzers.
It has itself an `analyzers` parameter that can be filled with other analyzers to group them.

> An example of this is also in (see [example_analyzers.yaml](diagnostic_aggregator/example/example_analyzers.yaml)):

Reviewer: ```suggestion
An example of this is (see [example_analyzers.yaml](diagnostic_aggregator/example/example_analyzers.yaml)):
```
